### ISSUE TITLE
[MNG-6495] ModelResolver cannot be null

### ISSUE THREAD

**Issue by jira-importer** [apache/maven#7370]
**[Elliotte Rusty Harold](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=elharo)** opened **[MNG-6495](https://issues.apache.org/jira/browse/MNG-6495?redirect=false)** and commented

Got this exception stacktrace while writing some of my own code today:

```
Exception in thread "main" java.lang.NullPointerException: request.modelResolver cannot be null (parent POM com.google.guava:guava-parent:26.0-jre and POM com.google.guava:guava:[unknown-version])
	at org.apache.commons.lang3.Validate.notNull(Validate.java:225)
	at org.apache.maven.model.building.DefaultModelBuilder.readParentExternally(DefaultModelBuilder.java:1046)
	at org.apache.maven.model.building.DefaultModelBuilder.readParent(DefaultModelBuilder.java:830)
	at org.apache.maven.model.building.DefaultModelBuilder.build(DefaultModelBuilder.java:332)
	at com.google.cloud.tools.opensource.dependencies.MetadataExplorer.main(MetadataExplorer.java:50)
```

No big deal except that the JavaDoc at [ModelBuildingRequest](http://maven.apache.org/ref/3.5.4/maven-model-builder/apidocs/org/apache/maven/model/building/ModelBuildingRequest.html) says "The model resolver to use, may be null."

Not sure whether to update the docs or the code here.

---

**Affects:** 3.5.4

0 votes, 6 watchers

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Ouch...[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte), this an easy fix if we know which source is correct: Javadoc or source code. Ideas?

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

I don't understand the `[unknown-version]`-part. As if the parent-pom is invalid. Do you have a minimal pom to reproduce this issue?

**Comment by jira-importer** [comment]
**[Elliotte Rusty Harold](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=elharo)** commented

I'll have to see if I can repro this. This as not a command line invocation of Maven. It happened because some of my own code attempted to pass null for an argument.

**Comment by jira-importer** [comment]
**[Elliotte Rusty Harold](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=elharo)** commented

I think this is the code where I stumbled onto this. Looks like I since added a ModelResolver to avoid the issue, but try setting that to null instead and see what happens.

```
package com.google.cloud.tools.opensource.dependencies;

import java.net.MalformedURLException;
import java.net.URL;

import org.apache.maven.building.UrlSource;
import org.apache.maven.model.Model;
import org.apache.maven.model.building.DefaultModelBuilder;
import org.apache.maven.model.building.DefaultModelBuilderFactory;
import org.apache.maven.model.building.DefaultModelBuildingRequest;
import org.apache.maven.model.building.ModelBuildingException;
import org.apache.maven.model.building.ModelBuildingRequest;
import org.apache.maven.model.building.ModelBuildingResult;
import org.apache.maven.model.building.ModelSource;
import org.apache.maven.model.building.UrlModelSource;
import org.apache.maven.model.resolution.ModelResolver;
import org.apache.maven.model.resolution.UnresolvableModelException;
import org.apache.maven.project.ProjectModelResolver;
import org.eclipse.aether.artifact.DefaultArtifact;
import org.eclipse.aether.collection.DependencyCollectionException;
import org.eclipse.aether.resolution.DependencyResolutionException;

public class MetadataExplorer {

  public static void main(String[] args)
      throws DependencyCollectionException, DependencyResolutionException, MalformedURLException, UnresolvableModelException {
    
    try {
      DefaultArtifact artifact = new DefaultArtifact("com.google.guava:guava:26.0-jre");
      DefaultModelBuilder modelBuilder = new DefaultModelBuilderFactory().newInstance();

      ModelBuildingRequest request = new DefaultModelBuildingRequest();
      URL pomUrl = new URL(
          "https://repo1.maven.org/maven2/com/google/guava/guava/26.0-jre/guava-26.0-jre.pom");
      UrlModelSource source = new UrlModelSource(pomUrl);
      request.setModelSource(source);
      
      ProjectModelResolver modelResolver
          = new ProjectModelResolver(null, null, null, null, null, null, null);
      
      ModelSource s = modelResolver.resolveModel("com.google.guava", "guava", "26.0-jre");
      
      request.setModelResolver(modelResolver);
      ModelBuildingResult result = modelBuilder.build(request);
      Model model = result.getEffectiveModel();
      System.out.println(model.getArtifactId());
    } catch (IllegalArgumentException ex) {
      System.err.println("Bad Maven coordinates " + args[0]);
      return;      
    } catch (ModelBuildingException e) {
      // TODO Auto-generated catch block
      e.printStackTrace();
    }
  }

}
```

**Comment by jira-importer** [comment]
**[Sylwester Lachiewicz](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=slachiewicz)** commented

Looks like we need to update doc here: [ModelBuildingRequest:298](https://gitbox.apache.org/repos/asf?p=maven.git;a=blob;f=maven-model-builder/src/main/java/org/apache/maven/model/building/ModelBuildingRequest.java;h=bc6e71bee21c2c97538a89bad931dc34ad1574cc;hb=HEAD#l298) ?

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Fixed with [c674bcfb426be6425471c541769b161ccef84586](https://gitbox.apache.org/repos/asf?p=maven-archiver.git;a=commit;h=c674bcfb426be6425471c541769b161ccef84586).
